### ISSUE TITLE
duration formatting

### ISSUE THREAD

**Issue by randomness666** [randomness666/test#26]
<a href="https://github.com/mj1856"><img src="https://avatars.githubusercontent.com/u/1396388?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [mj1856](https://github.com/mj1856)**
_Wednesday Sep 04, 2013 at 22:51 GMT_
_Originally opened as https://github.com/moment/moment/issues/1048_

----

We've touched on this before in #463 and #879 but never really talked about it directly in its own issue.

We need a method for formatting durations similar to how we format dates.   It should be as simple as:

```javascript
moment.duration(x).format("H:mm:ss")
```

Note that the formatting tokens would have to carry slightly different meaning, since we're representing an elapsed duration of time, rather than a time-of-day.  I suggest the following:

- `hh` would mean "hours remainder after accounting for days"
- `h` would be the single digit form of `hh`
- `HH` would mean "total whole hours"
- `H` would be the single digit form of `HH`
- `HHH` would mean "total hours including decimals" - although that can currently be done with `duration.asHours()`, so this might not be necessary.

Similar formatting would apply for other units.  In this context, the highest unit would be a "day", which would be a standard day consisting of 24 standard hours.  It wouldn't make sense to measure years or months with this due to calendaring issues.

Note, this came up recently (again) on StackOverflow in [this question](http://stackoverflow.com/q/18623783/634824).  The user was looking for something like `duration.format("H:mm:ss")`.  This is the workaround:

    Math.floor(duration.asHours()) + moment.utc(duration.asMilliseconds()).format(":mm:ss")

This works, but it feels hacky.  This should be built in.

**Comment by randomness666** [comment]
<a href="https://github.com/sbruchmann"><img src="https://avatars.githubusercontent.com/u/141435?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sbruchmann](https://github.com/sbruchmann)**
_Wednesday Sep 04, 2013 at 22:57 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/leo-mck"><img src="https://avatars.githubusercontent.com/u/876986?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [leo-mck](https://github.com/leo-mck)**
_Thursday Sep 05, 2013 at 11:56 GMT_

----

`moment.duration(x).format("H:mm:ss")` was exactly what I was expecting, before understanding that it would not work. The workaround suggested by Matt worked fine but comparing with the elegance of other things in momentjs it sure seems hacky.

**Comment by randomness666** [comment]
<a href="https://github.com/robinvdvleuten"><img src="https://avatars.githubusercontent.com/u/238295?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [robinvdvleuten](https://github.com/robinvdvleuten)**
_Saturday Sep 07, 2013 at 15:00 GMT_

----

Isn't it better to do
```
moment.utc(total.asMilliseconds()).format("HH:mm:ss");
```

instead of
```
Math.floor(duration.asHours()) + moment.utc(duration.asMilliseconds()).format(":mm:ss")
```

**Comment by randomness666** [comment]
<a href="https://github.com/ghost"><img src="https://avatars.githubusercontent.com/u/10137?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ghost](https://github.com/ghost)**
_Saturday Sep 07, 2013 at 20:32 GMT_

----

[![Code Bounty](https://app.codebounty.co/reward/image/f31fccc6b412ddd8b6068474)](https://app.codebounty.co/reward/link/f31fccc6b412ddd8b6068474)

**Comment by randomness666** [comment]
<a href="https://github.com/mj1856"><img src="https://avatars.githubusercontent.com/u/1396388?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mj1856](https://github.com/mj1856)**
_Sunday Sep 08, 2013 at 02:10 GMT_

----

@RobinvdVleuten  - The problem with using the `HH` in the current formatter is that it represents the hour *portion* of a full date and time.  When dealing with an *elapsed* time, there could well be 24 hours or more.  Try passing 24 hours in, and you'll see that it gets reset to zero.

**Comment by randomness666** [comment]
<a href="https://github.com/robinvdvleuten"><img src="https://avatars.githubusercontent.com/u/238295?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [robinvdvleuten](https://github.com/robinvdvleuten)**
_Wednesday Sep 11, 2013 at 08:44 GMT_

----

I noticed it indeed. I'm now using the calculated milliseconds to format it, but it's still a hacky workaround.

**Comment by randomness666** [comment]
<a href="https://github.com/icambron"><img src="https://avatars.githubusercontent.com/u/244988?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [icambron](https://github.com/icambron)**
_Monday Sep 16, 2013 at 07:15 GMT_

----

Are you thinking there will just be one "level" of remaindering possible--e.g. no way to say "hours after accounting for months"? I think that's a perfectly reasonable limitation, but just to spitball a sec, I could imagine providing an "anchor" argument, like so:

```js
moment.duration(x).format("hh:mm:ss", "M");
```

Which tells us that the hours should be modulo the month. The anchor is optional and would default to the largest unit plus one size up (in this case, days). Similarly, it could use the string itself to determine that the minutes are modulo the hours and seconds modulo minutes. Whereas if you did this:

```js
moment.duration(x).format("hh:ss");
```

...the seconds would be modulo the hours.

That's probably all not a great idea--it's not ironed out (where to weeks go in this hierarchy?) and it's probably unnecessary (people are unlikely to want holes like that). But fun to think about.

**Comment by randomness666** [comment]
<a href="https://github.com/icambron"><img src="https://avatars.githubusercontent.com/u/244988?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [icambron](https://github.com/icambron)**
_Monday Sep 16, 2013 at 07:17 GMT_

----

On a more serious note, how are you going to deal with month vs minutes with this lowercase/uppercase notation?

**Comment by randomness666** [comment]
<a href="https://github.com/mj1856"><img src="https://avatars.githubusercontent.com/u/1396388?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mj1856](https://github.com/mj1856)**
_Monday Sep 16, 2013 at 18:59 GMT_

----

I was envisioning that days would be the maximum unit size, so months wouldn't enter the picture.  Basically, this is the inverse of how we parse durations coming from asp.net time spans.

**Comment by randomness666** [comment]
<a href="https://github.com/icambron"><img src="https://avatars.githubusercontent.com/u/244988?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [icambron](https://github.com/icambron)**
_Monday Sep 16, 2013 at 19:09 GMT_

----

Makes sense.

**Comment by randomness666** [comment]
<a href="https://github.com/lfnavess"><img src="https://avatars.githubusercontent.com/u/3844871?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [lfnavess](https://github.com/lfnavess)**
_Thursday Oct 31, 2013 at 22:30 GMT_

----

I have been reading a SCORM 2004 RTE (p77) document, and i found that for time interval representation is as follows, i think this came from a standar,

in theory i can pass a value like: P34H
and the parsers should interpret that as 1day and 10 Hours

i hope it could be posible to format the ouput with this in mind
like 'PHM'
and the formatter interpret that as total Hours remaining minutes?
i would like to be able to format the output like HH:MM


i hope this info is helpfull

timeinterval (second, 10,2):The timeinterval (second, 10, 2)denotes that the value for 
the data model element timeinterval represents elapsed time with a precision of 0.01 
seconds[1]. The SCORM dot-notation binding defines a particular format for a 
characterstring to represent a timeinterval. 
The format of the characterstring shall be as follows: 
P[yY][mM][dD][T[hH][nM][s[.s]S]] where: 
•  y: The number of years (integer, >= 0, not restricted) 
•  m: The number of months (integer, >=0, not restricted) 
•  d: The number of days (integer, >=0, not restricted) 
•  h: The number of hours (integer, >=0, not restricted) 
•  n: The number of minutes (integer, >=0, not restricted) 
•  s: The number of seconds or fraction ofseconds (real or integer, >=0, not 
restricted). If fractions of a second are used, SCORM further restricts the string to 
a maximum of 2 digits (e.g., 34.45 – valid, 34.45454545 – not valid). 
•  The character literals designators P, Y, M, D, T, H, Mand Sshall appear if the 
corresponding non-zero value is present. 
•  Zero-padding of the values shall be supported. Zero-padding does not change the 
integer value of the number being represented by a set of characters. For 
example, PT05H is equivalent to PT5H and PT000005H. 
Example: 
•  P1Y3M2DT3H indicates a period of time of 1 year, 3 months, 2 days and 3 hours 
•  PT3H5M indicates a period of time of 3 hours and 5 minutes 

Implementers should be aware that the format and binding is for the communication of 
the data between a SCO and an LMS. Since the format is representing a period of time, 
then a duration like PT5M is equivalent to PT300S. 
If the data model element, that is of type timeinterval(second,10,2) contains a value, then 
•  The designator Pshall be present; 
•  If the value of years, months, days, hours, minutes or seconds is zero, the value 
and corresponding character literal designation may be omitted, but at least one 
character literal designator and value shall be present in addition to the designator 
P; 
•  The designator Tshall be omitted if all of the time components (hours, minutes 
and seconds) are not used. A zero value may be used with any of the time 
components (e.g., PT0S).

**Comment by randomness666** [comment]
<a href="https://github.com/quietmint"><img src="https://avatars.githubusercontent.com/u/622385?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quietmint](https://github.com/quietmint)**
_Tuesday Nov 26, 2013 at 21:53 GMT_

----

Duration formatting would be an excellent addition. There's no need for two distinct "hour" tokens. Assume a duration of 32 hours. You'd never want to extract just 8 hours without also extracting 1 day.

I think @icambron has a good suggestion. Parse the format string into tokens and express the duration in terms of the largest unit. Parse the tokens one at a time in order (e.g., "DD:hh:ss" meaning number of complete days, number of complete hours after accounting for days, number of complete seconds after accounting for days + hours).

**Comment by randomness666** [comment]
<a href="https://github.com/jsmreese"><img src="https://avatars.githubusercontent.com/u/1275528?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jsmreese](https://github.com/jsmreese)**
_Saturday Dec 28, 2013 at 00:37 GMT_

----

I've posted a moment.duration.format plugin:
https://github.com/jsmreese/moment-duration-format

I think it addresses most of the ideas / use-cases in this thread.

**Comment by randomness666** [comment]
<a href="https://github.com/hotzenklotz"><img src="https://avatars.githubusercontent.com/u/1105056?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hotzenklotz](https://github.com/hotzenklotz)**
_Thursday Mar 13, 2014 at 04:10 GMT_

----

@jsmreese Have you consider filing a pull request that contains your plugin as a core part of momentjs?

**Comment by randomness666** [comment]
<a href="https://github.com/jsmreese"><img src="https://avatars.githubusercontent.com/u/1275528?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jsmreese](https://github.com/jsmreese)**
_Wednesday Mar 19, 2014 at 16:18 GMT_

----

@hotzenklotz Yes I have considered filing a pull request.

I haven't done so because of all the reasons @icambron laid out in #1538.

My plugin:
- depends on Lo-Dash
- does not look and feel like Moment.js code
- uses a completely different test setup

I would *love* for my plugin to become part of Moment.js core... but I'm not going to waste their time with a pull request before those issues are addressed.

**Comment by randomness666** [comment]
<a href="https://github.com/schmod"><img src="https://avatars.githubusercontent.com/u/1102667?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schmod](https://github.com/schmod)**
_Wednesday Mar 19, 2014 at 16:52 GMT_

----

We would also want to make sure that the code can be internationalized.  Fortunately, most of the strings we would need are in the Unicode CLDR, so very little translation work should be necessary.  

The CLDR also has locale-specific recommendations on how to format certain kinds of intervals, which could potentially be more useful than arbitrary duration formats.  Not sure how that fits in here, but it may be nice to automagically be able to display locale-specific intervals between two specific times.

The CLDR also has information about how to display specific time intervals (rather than durations) which could be useful at some point... 

[English(US) Calendar Data](https://github.com/jquery/globalize/blob/master/test/fixtures/cldr/main/en/ca-gregorian.json)

**Comment by randomness666** [comment]
<a href="https://github.com/jsmreese"><img src="https://avatars.githubusercontent.com/u/1275528?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jsmreese](https://github.com/jsmreese)**
_Monday Jul 28, 2014 at 05:34 GMT_

----

Just published a new version of Moment Duration Format that removes the previous dependency on Lo-Dash or Underscore.

https://github.com/jsmreese/moment-duration-format

**Comment by randomness666** [comment]
<a href="https://github.com/nathanhoel"><img src="https://avatars.githubusercontent.com/u/5025854?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nathanhoel](https://github.com/nathanhoel)**
_Thursday Aug 14, 2014 at 20:04 GMT_

----

@jsmreese Suits our needs perfectly. Thanks!

**Comment by randomness666** [comment]
<a href="https://github.com/outaTiME"><img src="https://avatars.githubusercontent.com/u/140472?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [outaTiME](https://github.com/outaTiME)**
_Monday Dec 29, 2014 at 14:59 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/mhayes14"><img src="https://avatars.githubusercontent.com/u/591318?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mhayes14](https://github.com/mhayes14)**
_Tuesday Jan 13, 2015 at 00:15 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/rsweetland"><img src="https://avatars.githubusercontent.com/u/81267?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [rsweetland](https://github.com/rsweetland)**
_Saturday Feb 07, 2015 at 00:00 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/larssondaniel"><img src="https://avatars.githubusercontent.com/u/1529573?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [larssondaniel](https://github.com/larssondaniel)**
_Saturday Feb 07, 2015 at 00:08 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/mlegenhausen"><img src="https://avatars.githubusercontent.com/u/424493?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mlegenhausen](https://github.com/mlegenhausen)**
_Wednesday Apr 29, 2015 at 11:56 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/oznu"><img src="https://avatars.githubusercontent.com/u/3979615?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [oznu](https://github.com/oznu)**
_Tuesday May 19, 2015 at 11:27 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/rhobot"><img src="https://avatars.githubusercontent.com/u/872220?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [rhobot](https://github.com/rhobot)**
_Friday May 22, 2015 at 21:02 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/diegoleme"><img src="https://avatars.githubusercontent.com/u/374857?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [diegoleme](https://github.com/diegoleme)**
_Wednesday May 27, 2015 at 21:24 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/schmod"><img src="https://avatars.githubusercontent.com/u/1102667?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schmod](https://github.com/schmod)**
_Thursday May 28, 2015 at 16:03 GMT_

----

I probably sound like a broken record on this issue, but I would like to ensure that any solution follows existing conventions wherever possible, and can easily be internationalized (which includes adding symbols similar to `LLLL` for commonly-used duration formats.

Somewhat annoyingly, the CLDR doesn't provide many specific guidelines for formatting durations, although there are extensive relative duration and interval guidelines which could be of use to moment elsewhere.  However, they do provide some minimal guidelines and translations for [duration units](https://github.com/unicode-cldr/cldr-units-full/blob/master/main/en-US-POSIX/units.json), which would allow you to concatenate the relevant units when implementing something like `humanize()`.

The `hh` / `h` syntax described above feels like a significant departure from the formatting tokens used by ISO8601, CLDR, and [Apache](http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/time/DurationFormatUtils.html), and I would prefer to avoid it if at all possible.  

A better proposal might be to infer to use the most significant unit in the formatting pattern as the modulo, so `h:mm:ss` would display a "normal" time, but count hours past 24 (eg 26:30:00).  It's unclear how one would even compute a pattern like `"HH:MM:ss"`, or what the use case for that would be.  Allowing developers to override this behavior also seems like it could easily become a source of bugs.

Keeping in the spirit of "i18n everywhere," the CLDR defines duration formats for:
* Hours + minutes ( `h:mm` in EN-US)
* Hours + minutes + second (`h:mm:ss` in EN-US)
* Minutes + seconds (`m:ss` in EN-US)
and it might make sense to provide locale-specific constants (similar to the `LLL` date formats) for these time durations.

Unfortunately, formatting durations with units greater than hours is really difficult to express via a single formatting string (thanks to the pluralization rules that you'd need to consider), and I wasn't able to find *any* library in any language that allows for easy, i18n-friendly formatting of durations longer than 24 hours.  The best that you can do would be to extend `duration.humanize()` to take some extra parameters, effectively implementing the original proposal in #463.  

In short, it might not be a good idea to implement `duration.format()`, as I see any potential solution having considerable pitfalls.  Your best bet would be to improve `duration.humanize()`, or implement a scaled-back `duration.format()` that only understands hours, minutes, and seconds.

**Comment by randomness666** [comment]
<a href="https://github.com/pracucci"><img src="https://avatars.githubusercontent.com/u/1701904?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [pracucci](https://github.com/pracucci)**
_Thursday May 28, 2015 at 16:12 GMT_

----

+1

In case you need a quick function to add padding:
```
function padWithZero(input, length) {
    // Cast input to string
    input = "" + input;

    let paddingSize = Math.max(0, length - input.length);
    return new Array(paddingSize > 0 ? paddingSize + 1 : 0).join("0") + input;
}
```

**Comment by randomness666** [comment]
<a href="https://github.com/bsr203"><img src="https://avatars.githubusercontent.com/u/1966521?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bsr203](https://github.com/bsr203)**
_Thursday Jun 04, 2015 at 20:29 GMT_

----

+1

**Comment by randomness666** [comment]
<a href="https://github.com/FourwingsY"><img src="https://avatars.githubusercontent.com/u/3432949?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [FourwingsY](https://github.com/FourwingsY)**
_Wednesday Jun 17, 2015 at 06:09 GMT_

----

+1
